manifest/package.go

```diff
@@ -90,6 +98,87 @@ func (r *manifestConverter) validateManifest(manif *Manifest) error {
+	packageModulesMapping := make(map[string]*pbsubstreams.Module)
+	for _, module := range pkg.Modules.Modules {
+		packageModulesMapping[module.Name] = module
+	}
+
+	for _, manifestModule := range manif.Modules {
+		if manifestModule.Use != "" {
+			usedModule, found := packageModulesMapping[manifestModule.Use]
+			if !found {
+				return fmt.Errorf("module %q: use module %q not found", manifestModule.Name, manifestModule.Use)
+			}
+			moduleWithUse := packageModulesMapping[manifestModule.Name]
+
+			if err := checkEqualInputs(moduleWithUse, usedModule, manifestModule, packageModulesMapping); err != nil {
+				return fmt.Errorf("checking inputs for module %q: %w", manifestModule.Name, err)
+			}
+
+			moduleWithUse.BinaryIndex = usedModule.BinaryIndex
+			moduleWithUse.BinaryEntrypoint = usedModule.BinaryEntrypoint
+
+			if moduleWithUse.InitialBlock == 0 {
+				moduleWithUse.InitialBlock = usedModule.InitialBlock
+			}
+
+			moduleWithUse.Output = usedModule.Output
+			moduleWithUse.Kind = usedModule.Kind
+		}
+	}
+	return nil
+}
+
+func checkEqualInputs(moduleWithUse, usedModule *pbsubstreams.Module, manifestModuleWithUse *Module, packageModulesMapping map[string]*pbsubstreams.Module) error {
+	for index, input := range moduleWithUse.Inputs {
+		usedModuleInput := usedModule.Inputs[index]
+		if input.GetParams() != nil {
+			if usedModuleInput.GetParams() == nil {
+				return fmt.Errorf("module %q: input %q is not a params type", manifestModuleWithUse.Name, input.String())
+			}
+			if input.GetParams().Value != usedModuleInput.GetParams().Value {
+				return fmt.Errorf("module %q: input %q has different value than the used module %q: input %q", manifestModuleWithUse.Name, input.String(), manifestModuleWithUse.Use, usedModuleInput.String())
+			}
+		}
+		if input.GetSource() != nil {
+			if usedModuleInput.GetSource() == nil {
+				return fmt.Errorf("module %q: input %q is not a source type", manifestModuleWithUse.Name, input.String())
+			}
+			if input.GetSource().Type != usedModuleInput.GetSource().Type {
+				return fmt.Errorf("module %q: input %q has different source than the used module %q: input %q", manifestModuleWithUse.Name, input.String(), manifestModuleWithUse.Use, usedModuleInput.String())
+			}
+		}
+		if input.GetStore() != nil {
+			if usedModuleInput.GetStore() == nil {
+				return fmt.Errorf("module %q: input %q is not a store type", manifestModuleWithUse.Name, input.String())
+			}
+			if input.GetStore().GetMode() != usedModuleInput.GetStore().GetMode() {
+				return fmt.Errorf("module %q: input %q has different mode than the used module %q: input %q", manifestModuleWithUse.Name, input.String(), manifestModuleWithUse.Use, usedModuleInput.String())
+			}
+
+			inputStoreModuleName := input.GetStore().ModuleName
+			usedModuleStoreMapModuleName := usedModuleInput.GetStore().ModuleName
+			if packageModulesMapping[inputStoreModuleName].Output.Type != packageModulesMapping[usedModuleStoreMapModuleName].Output.Type {
```

Comment:
watch out for panic here, we want to first check if that new module name exists and error out correctly.